pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -14,227 +14,202 @@
-    C2 = U[0:2, 2:4]
-    C3 = U[2:4, 0:2]
-    C4 = U[2:4, 2:4]
-
-    # From the definition of A \otimes B, C1 C4^\dag = a1^2 I, so we can extract a1
-    C14 = math.dot(C1, math.conj(math.T(C4)))
-    a1 = math.sqrt(C14[0, 0])
+    # First let's compute gamma(u). For the one-CNOT case, uuT is always real.
+    u = math.dot(math.cast_like(Edag, U), math.dot(swap_U, math.cast_like(E, U)))
+    uuT = math.dot(u, math.T(u))
 
-    # Similarly, -C2 C3^\dag = a2^2 I, so we can extract a2
-    C23 = math.dot(C2, math.conj(math.T(C3)))
-    a2 = math.sqrt(-C23[0, 0])
+    # Since uuT is real, we can use eigh of its real part. eigh also orders the
+    # eigenvalues in ascending order.
+    _, p = math.linalg.eigh(qml.math.real(uuT))
 
-    # This gets us a1, a2 up to a sign. To resolve the sign, ensure that
-    # C1 C2^dag = a1 a2* I
-    C12 = math.dot(C1, math.conj(math.T(C2)))
+    # Fix the determinant if necessary so that p is in SO(4)
+    if math.linalg.det(p) < 0:
+        p = math.dot(p, math.cast_like(LAST_COL_NEG, p))
 
-    if not math.allclose(a1 * np.conj(a2), C12[0, 0]):
-        a2 *= -1
+    # Now, we must find q such that p uu^T p^T = q vv^T q^T.
+    # For this case, our V = SWAP CNOT01 is constant. Thus, we can compute v,
+    # vvT, and its eigenvalues and eigenvectors directly. We store them here as
+    # constants.
+    v = np.array(
+        [
+            [0.5, 0.5j, 0.5j, -0.5],
+            [-0.5j, 0.5, -0.5, -0.5j],
+            [-0.5j, -0.5, 0.5, -0.5j],
+            [0.5, -0.5j, -0.5j, -0.5],
+        ]
+    )
+
+    # This q is properly in SO(4)
+    q = (1 / np.sqrt(2)) * np.array([[-1, 0, -1, 0], [0, 1, 0, 1], [0, 1, 0, -1], [1, 0, -1, 0]])
+
+    # Once we have p and q properly in SO(4), we compute G and H in SO(4) such
+    # that U = G V H
+    G = math.dot(p, q.T)
+    H = math.dot(math.conj(math.T(v)), math.dot(math.T(G), u))
 
-    # Construct A
-    A = math.stack([[a1, a2], [-math.conj(a2), math.conj(a1)]])
+    # We now use the magic basis to convert G, H to SU(2) x SU(2)
+    AB = math.dot(E, math.dot(G, Edag))
+    CD = math.dot(E, math.dot(H, Edag))
 
-    # Next, extract B. Can do from any of the C, just need to be careful in
-    # case one of the elements of A is 0.
-    if not math.allclose(A[0, 0], 0.0, atol=1e-6):
-        B = C1 / math.cast_like(A[0, 0], 1j)
-    else:
-        B = C2 / math.cast_like(A[0, 1], 1j)
+    # Extract the tensor prodcts to SU(2) x SU(2)
+    A, B = _su2su2_to_tensor_products(AB)
+    C, D = _su2su2_to_tensor_products(CD)
 
-    return math.convert_like(A, U), math.convert_like(B, U)
+    # Recover the operators in the decomposition; note that because of the
+    # initial SWAP, we exchange the order of A and B
+    A_ops = zyz_decomposition(A, wires[1])
+    B_ops = zyz_decomposition(B, wires[0])
+    C_ops = zyz_decomposition(C, wires[0])
+    D_ops = zyz_decomposition(D, wires[1])
 
+    return C_ops + D_ops + [qml.CNOT(wires=[wires[0], wires[1]])] + A_ops + B_ops
```

Comment:
very very minor, but does `qml.CNOT(wires=wires)` not work?

Reply:
```suggestion
    return C_ops + D_ops + [qml.CNOT(wires=wires)] + A_ops + B_ops
```